### PR TITLE
CURA-11978 Retract and unretract in a travel

### PR DESCRIPTION
**Pull request by wawanbreton**
CURA-11978

### CODE REVIEW

File: src/gcodeExport.cpp [hunk 1]
```diff
@@ -1249,48 +1300,80 @@ void GCodeExport::writeRetraction(const RetractionConfig& config, bool force, bo
...
+
+    RetractionAmounts retraction_amounts;
+    if (retract_distance.has_value())
+    {
+        retraction_amounts = computeRetractionAmounts(extruder_attr_[current_extruder_], *retract_distance);
+    }
+
+    is_z_hopped_ = height;
+    const coord_t target_z = current_layer_z_ + is_z_hopped_;
+    // current_position_.z_ = current_layer_z_;
```
Comment (jellespijker) [inline src/gcodeExport.cpp:1364]: Commented out code, please remove

File: src/LayerPlan.cpp [hunk 2]
```diff
@@ -2152,6 +2153,238 @@ void LayerPlan::addLinesInGivenOrder(
...
+        const Ratio retract_prime_ratio = retract_duration_during_travel / total_anti_ooze_duration_during_travel;
+
+        const Duration allowed_retract_duration_during_travel = travel_durations.travel * retract_prime_ratio;
+        extra_time_still_retraction = retract_duration_during_travel - allowed_retract_duration_during_travel;
+
+        const Duration allowed_prime_duration_during_travel = travel_durations.travel * (1.0_r - retract_prime_ratio);
+        extra_time_still_prime = prime_duration_during_travel - allowed_prime_duration_during_travel;
+    }
+
+    const auto compute_anti_ooze_amounts
```
Comment (jellespijker) [inline src/LayerPlan.cpp:2225]: If would make this an actual function.This would not only improve readability but also make it easier to write unit tests against the logic for computing the per‑segment amounts.

File: include/gcodeExport.h [hunk 3]
```diff
@@ -113,6 +115,22 @@ class GCodeExport : public NoCopy
...
+        double new_e{ 0.0 }; // The new absolute retraction amount
+        double diff_e{ 0.0 }; // The difference between the new and previous amount, which is to be processed
+
+        /*!
+         * Indicates whether this retraction actually has something to process
+         */
+        inline bool has_retraction() const
+        {
+            return std::abs(diff_e) >= 0.000001;
+        }
```
Comment (jellespijker) [inline include/gcodeExport.h:131]: `nodiscard` is your friend and member functions defined inside a class definition are implicitly inline.The primary improvements come from making the function `constexpr` and `noexcept`.

```suggestion
        [[nodiscard]] constexpr bool has_retraction() const noexcept
        {
            constexpr double threshold { 1e-6 };
            return std::abs(diff_e) >= threshold;
        }
```

File: include/utils/types/numeric_facade.h [hunk 4]
```diff
@@ -14,7 +14,7 @@ struct NumericFacade
 {
     using value_type = T;
 
-    value_type value{};
+    value_type value{ static_cast<T>(0) };
```
Comment (jellespijker) [inline include/utils/types/numeric_facade.h:17]: not sure if a static_cast is needed here, for simple numerical types, the default is always a zero value
```suggestion
    value_type value{ };
```

File: src/LayerPlan.cpp [hunk 5]
```diff
@@ -2152,6 +2153,238 @@ void LayerPlan::addLinesInGivenOrder(
...
+        const Velocity& travel_speed = path.config.getSpeed();
+        Point2LL start_position = gcode.getPosition().toPoint2LL();
+        for (const Point3LL& travel_point : path.points)
+        {
+            const Point2LL travel_point_2d = travel_point.toPoint2LL();
+            const coord_t travel_segment_length = vSize(travel_point_2d - start_position);
+            travel_durations.travel += (travel_segment_length / travel_speed) / 1000.0;
+            start_position = travel_point_2d;
+        }
+    }
```
Comment (jellespijker) [inline src/LayerPlan.cpp:2177]: Untested and a suggestion, it should be a bit more performant avoids some extra calculations
```suggestion
    if (! path.points.empty())
    {
        const Velocity travel_speed = path.config.getSpeed();
        const double factor = 1.0 / travel_speed / 1000.0;
        Point2LL start = gcode.getPosition().toPoint2LL();
        auto total_distance = std::accumulate(
            path.points.begin(), path.points.end(),
            0.0,
            [&start](double acc, const Point3LL& pt)
            {
                const Point2LL current = pt.toPoint2LL();
                double seg = vSize(current - start);
                start = current;
                return acc + seg;
            });
        travel_durations.travel += total_distance * factor;
    }
```

**Review by jellespijker — CHANGES_REQUESTED** [review]
Overal it looks good and well thought out, a couple of minor remarks and suggestions. But I do suggest adding some UT
given the complexity of splitting travel moves into multiple phases, maybe add unit and integration tests that cover:

- A travel move with enough duration for all retraction/priming to happen during travel.
- A travel move where extra stationary time is required.
- Edge cases with only one travel segment.
- Both “forward” (retraction) and “reversed” (priming) computations.

File: include/utils/types/numeric_facade.h (hunk 4 above)
Comment (wawanbreton) [reply include/utils/types/numeric_facade.h:17]: Hem, unless I missed some improvement of the language, it is not. Basic types like `float` and `int` don't have an initial value and can contain anything. Or maybe just having the `{}` is good enough ?

Comment (wawanbreton) [reply include/utils/types/numeric_facade.h:17]: Ok, seems to be the case indeed. I thought `value_type value{};` and `value_type value;` were identical but they are not.

Comment (jellespijker) [reply include/utils/types/numeric_facade.h]: https://en.cppreference.com/w/cpp/language/zero_initialization

File: include/gcodeExport.h (hunk 3 above)
Comment (jellespijker) [reply include/gcodeExport.h]: Didn't you pay attention `constexpr` everything! https://www.youtube.com/watch?v=PJwd4JLYJJY

Comment (jellespijker) [reply include/gcodeExport.h]: But yes in this case it is still depend on a dynamic value

**Review by jellespijker — APPROVED** [review]
LGTM

File: Cura.proto [hunk 6]
```diff
@@ -83,10 +83,12 @@ message Polygon {
...
         InfillType = 6;
         SupportInfillType = 7;
-        MoveCombingType = 8;
-        MoveRetractionType = 9;
+        MoveUnretracted = 8;
+        MoveRetracted = 9;
         SupportInterfaceType = 10;
         PrimeTowerType = 11;
+        MoveWhileRetracting = 12;
+        MoveWhileUnretracting = 13;
```
Comment (HellAholic) [inline Cura.proto:91]: Should we add the
    StationaryRetractUnretract = 14;
    NumPrintFeatureTypes = 15;
?

Comment (wawanbreton) [reply Cura.proto:91]: Oh, definitely... I dont't see how it can work without it actually :thinking:
